src/buildercore/bootstrap.py

```diff
@@ -429,6 +429,32 @@ def download_master_builder_key(stackname):
         operations.get(remote_path=private_key, local_path=fh, use_sudo=True)
     return fh
 
+def download_master_configuration(master_stack):
+    fh = StringIO()
+    with stack_conn(master_stack):
+        operations.get(remote_path='/etc/salt/master.template', local_path=fh, use_sudo=True)
+    return fh
+
+def render_master_configuration(master_configuration_template, formulas):
+    cfg = core_utils.ordered_load(master_configuration_template)
+
+    def basename(formula):
+        return re.sub('-formula$', '', os.path.basename(formula))
+    formula_path = '/opt/formulas/%s/'
+
+    cfg['file_roots']['base'] = ["/opt/builder-private/salt/"] + [formula_path % basename(f) for f in formulas] + ["/opt/formulas/builder-base/"]
+    cfg['pillar_roots']['base'] = ["/opt/builder-private/pillar"]
+    # dealt with at the infrastructural level
+    cfg['interface'] = '0.0.0.0'
+
+    master_configuration = StringIO()
+    core_utils.ordered_dump(cfg, master_configuration)
+    return master_configuration
```

Comment:
another instance of a file-like object being returned instead of a string. ignore if intentional.